File: sample/ruby/17_monte_carlo_pi.rb
require_relative "py_runtime"


# 17: Sample that scans a large grid using integer arithmetic only and computes a checksum.
# It avoids floating-point error effects, making cross-language comparisons easier.

def run_integer_grid_checksum(width, height, seed)
  mod_main = 2147483647
  mod_out = 1000000007
  acc = (seed % mod_out)
  __step_0 = __pytra_int(1)
  y = __pytra_int(0)
  while ((__step_0 >= 0 && y < __pytra_int(height)) || (__step_0 < 0 && y > __pytra_int(height)))
    row_sum = 0
    __step_1 = __pytra_int(1)
    x = __pytra_int(0)
    while ((__step_1 >= 0 && x < __pytra_int(width)) || (__step_1 < 0 && x > __pytra_int(width)))
      v = ((((x * 37) + (y * 73)) + seed) % mod_main)
      v = (((v * 48271) + 1) % mod_main)
      row_sum += (v % 256)
      x += __step_1
    end
    acc = ((acc + (row_sum * (y + 1))) % mod_out)
    y += __step_0
  end
  return acc
end

def run_integer_benchmark()
  width = 7600
  height = 5000
  start = __pytra_perf_counter()
  checksum = run_integer_grid_checksum(width, height, 123456789)
  elapsed = (__pytra_perf_counter() - start)
  __pytra_print("pixels:", (width * height))
  __pytra_print("checksum:", checksum)
  __pytra_print("elapsed_sec:", elapsed)
end

if __FILE__ == $PROGRAM_NAME
  run_integer_benchmark()
end

File: sample/kotlin/17_monte_carlo_pi.kt
import kotlin.math.*


// 17: Sample that scans a large grid using integer arithmetic only and computes a checksum.
// It avoids floating-point error effects, making cross-language comparisons easier.

fun run_integer_grid_checksum(width: Long, height: Long, seed: Long): Long {
    var mod_main: Long = 2147483647L
    var mod_out: Long = 1000000007L
    var acc: Long = (seed % mod_out)
    var y = __pytra_int(0L)
    while (y < __pytra_int(height)) {
        var row_sum: Long = 0L
        var x = __pytra_int(0L)
        while (x < __pytra_int(width)) {
            var v: Long = ((((x * 37L) + (y * 73L)) + seed) % mod_main)
            v = (((v * 48271L) + 1L) % mod_main)
            row_sum += (v % 256L)
            x += 1L
        }
        acc = ((acc + (row_sum * (y + 1L))) % mod_out)
        y += 1L
    }
    return acc
}

fun run_integer_benchmark() {
    var width: Long = 7600L
    var height: Long = 5000L
    var start: Double = __pytra_perf_counter()
    var checksum: Long = __pytra_int(run_integer_grid_checksum(width, height, 123456789L))
    var elapsed: Double = (__pytra_perf_counter() - start)
    __pytra_print("pixels:", (width * height))
    __pytra_print("checksum:", checksum)
    __pytra_print("elapsed_sec:", elapsed)
}

fun main(args: Array<String>) {
    run_integer_benchmark()
}
